src/Microsoft.OData.Client/Metadata/ODataTypeInfo.cs

```diff
@@ -321,12 +321,18 @@ private PropertyInfo[] GetKeyProperties()
                     throw Client.Error.InvalidOperation(Error.Format(SRResources.ClientType_KeysOnDifferentDeclaredType, typeName));
                 }
 
-                // Check if the key property's type is a known primitive, an enum, or a nullable generic.
+                // Check if the key property's type is a known primitive, or an enum.
                 // If it doesn't meet any of these conditions, throw an InvalidOperationException.
-                if (!PrimitiveType.IsKnownType(key.PropertyType) && !key.PropertyType.IsEnum() && !(key.PropertyType.IsGenericType() && key.PropertyType.GetGenericTypeDefinition() == typeof(System.Nullable<>) && key.PropertyType.GetGenericArguments().First().IsEnum()))
+                if (!PrimitiveType.IsKnownType(key.PropertyType) && !key.PropertyType.IsEnum() && !(key.PropertyType.IsGenericType() && key.PropertyType.GetGenericArguments().First().IsEnum()))

```

Comment:
@WanjohiSammy  I think we can avoid using LINQ `First()` here - use `key.PropertyType.GetGenericArguments()[0]`